use-location.js

```diff
@@ -17,55 +21,51 @@ const eventPushState = "pushState";
 const eventReplaceState = "replaceState";
 export const events = [eventPopstate, eventPushState, eventReplaceState];
 
-export default ({ base = "" } = {}) => {
-  const [{ path }, update] = useState(() => ({ path: relativePath(base) }));
-  // @see https://reactjs.org/docs/hooks-reference.html#lazy-initial-state
-  const prevHash = useRef(path + location.search);
-
-  useEffect(() => {
-    // this function checks if the location has been changed since the
-    // last render and updates the state only when needed.
-    // unfortunately, we can't rely on `path` value here, since it can be stale,
-    // that's why we store the last pathname in a ref.
-    const checkForUpdates = () => {
-      const pathname = relativePath(base);
-      const hash = pathname + location.search;
-
-      if (prevHash.current !== hash) {
-        prevHash.current = hash;
-        update({ path: pathname });
-      }
-    };
-
-    events.forEach((e) => addEventListener(e, checkForUpdates));
+export const subscribeToLocationUpdates = (callback) => {
+  for (const event of events) {
+    window.addEventListener(event, callback);
+  }
+  return () => {
+    for (const event of events) {
+      window.removeEventListener(event, callback);
+    }
+  };
+};
 
-    // it's possible that an update has occurred between render and the effect handler,
-    // so we run additional check on mount to catch these updates. Based on:
-    // https://gist.github.com/bvaughn/e25397f70e8c65b0ae0d7c90b731b189
-    checkForUpdates();
+const currentSearch = () => location.search;
+export const useSearch = () =>
+  useSyncExternalStore(subscribeToLocationUpdates, currentSearch);
 
-    return () => events.forEach((e) => removeEventListener(e, checkForUpdates));
-  }, [base]);
+export const usePathname = (opts = {}) =>
+  useSyncExternalStore(subscribeToLocationUpdates, () =>
+    relativePath(opts.base || "")
+  );
```

Comment:
Can we keep the current way of using default options?

```suggestion
export const usePathname = ({ base = "" } = {}) =>
  useSyncExternalStore(subscribeToLocationUpdates, () => relativePath(base));
```

I know that your version handles cases like `base = false` or `null` but I don't think we should allow users to pass that as an option.

Reply:
@molefrog I was **not** actually doing that to support false or null, but rather to support `usePathname(router)` (since `base` may change inside that object, so it is important to look up the current value of base inside that object rather than copying the initial value supplied.

However... the option is actually a bit unsound in the first place, since `useSyncExternalStore` won't respond to base changes at all, thus if the base changes but the path does not, `usePathname(router)` won't work as expected. So I am going to modify this behavior to remove that option altogether from `usePathname` to limit the surface area of that behavior.